### PR TITLE
Streamed responses

### PR DESCRIPTION
**Pull request by samuelcolvin**
Turns out implementing streamed responses is a lot of work, but very worthwhile.

TODO:
* [x] Gemini support
* [x] check usage with logfire
* [x] docs for `whale.py`
* [x] test result validation
* [x] add streamed message to `result.all_messages`

### CODE REVIEW

File: pydantic_ai/_utils.py [hunk 1]
```diff
@@ -107,3 +109,65 @@ def is_left(self) -> bool:
...
+            wait_time = soft_max_interval - (time.monotonic() - group_start_time)
+
+        # if there's no current coroutine, we get the next one
+        if coro is None:
+            # aiter.__anext__() returns an Awaitable[T], not a Coroutine which asyncio.create_task expects
+            # TODO does this matter? It seems to run fine
+            coro = asyncio.create_task(aiter.__anext__())  # pyright: ignore[reportArgumentType]
+
+        # we use asyncio.wait to avoid cancelling the coroutine if it's not done
+        done, _ = await asyncio.wait((coro,), timeout=wait_time)
```
Comment (Viicos) [inline pydantic_ai/_utils.py]: In theory, it's possible for async iterators (described by the `AsyncIterator` ABC) to yield awaitable objects that are not coroutines and thus `create_task` will raise an exception at runtime. However, I think most implementations use the `async def` syntax, essentially creating a coroutine. If this utility function is private, it is probably going to be fine.

File: pydantic_ai/_utils.py [hunk 2]
```diff
@@ -107,3 +109,65 @@ def is_left(self) -> bool:
...
+            wait_time = soft_max_interval
+        else:
+            # wait for the time remaining in the group
+            wait_time = soft_max_interval - (time.monotonic() - group_start_time)
+
+        # if there's no current coroutine, we get the next one
+        if coro is None:
+            # aiter.__anext__() returns an Awaitable[T], not a Coroutine which asyncio.create_task expects
+            # TODO does this matter? It seems to run fine
+            coro = asyncio.create_task(aiter.__anext__())  # pyright: ignore[reportArgumentType]
```
Comment (Viicos) [inline pydantic_ai/_utils.py]:
```suggestion
            coro = asyncio.create_task(anext(aiter))  # pyright: ignore[reportArgumentType]
```
the built-in can be used as well

File: pydantic_ai/_utils.py (hunk 1 above)
Comment (Viicos) [reply pydantic_ai/_utils.py]: I've raised https://github.com/encode/httpx/discussions/3393

File: pydantic_ai/_utils.py (hunk 2 above)
Comment (samuelcolvin) [reply pydantic_ai/_utils.py]: not supported in 3.9 😞

File: pydantic_ai/_utils.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic_ai/_utils.py]: makes sense, I think the thing we're iterating over (at least for OpenAI) is not directly a httpx connection, but one from the openai library.

**Pull request merged**